### ISSUE TITLE
Multi-class detection with current frame work

### ISSUE THREAD

**Issue by placentian** [TensorBox/TensorBox#210]
Hi @Russell91, it is really nice to publish this great framework for object detection. It looks like not difficult at all to use it for other type of object detection with appropriate training data set. 

I am a little bit interested in multi-class detection recently and I know this is a very popular topic and have done all the due diligence to look through the previous discussions. I have some quick thoughts and would like to get your advice while I am working on the implementation. 

1. set the right labels (flags) based on training data to make the flag as the index of detection classes
2. In build_forward_backward(), update the labels for two sparse_softmax_cross_entropy_with_logits() calls. It looks like the labels only has the value 0 and 1 but it needs to be updated to the index of detection classes for the sparse_softmax_cross_entropy call. 
3. update add_rectangle function to add an extra loop to go through each class. 
    3.1 Current implementation consolidate the con (conf = np.max(confidences_r[0, y, x, n, 1:])). Do we need to take care of the bbox merging (stitching) code to only merge the bbox for the same class? or maybe that does not matter as only the box with the highest confidence will be selected. 
    3.2 add a class id for each merged box to differentiate the detection results for different classes. 

any response will be greatly appreciated.

**Comment by twhitloc** [comment]
Hi @placentian, I am also looking into it.  It seems that the right labels get lost in build_forward_backward() due to the tf.greater() call.  This returns a truth value about the comparison, and not the greater value.  Removing the call ensures that the correct label is maintained.  

I am still looking around, but as per your 2nd point that is where the value is lost.

**Comment by placentian** [comment]
HI @twhitloc  you are absolutely right. I guess the change here should be very easy. The input to this function is already the index of the classes and a minor change is needed. 

What do you think of the 3rd point or anything I missed?

**Comment by twhitloc** [comment]
I looked a little bit into modifying stitch_rects and hungarian and I think it is probably a good idea to include classIDs in these fields as there may be two overlapping entities of different classes.  

I tinkered with it last week but that was not including other changes that would ensure it worked properly.  I would expect a little bit of a performance (in terms of time) drop unfortunately.

**Comment by RomanSteinberg** [comment]
I want to mention that `confidences` variable is actuallly `box_flags` from `data_utils`. This means that it just points whether this cell contains any box or not. 

So, I think adding classID is not just easy task.

**Comment by placentian** [comment]
@RomanSteinberg  and @twhitloc ,Sorry for the confusion in my previous thread. I meant to use some variable (say class id) rather than classID field in rect to indicate different classes in detection (eval or test run). 

 I did some quick dirty work in add_rectangle (using different colors) and it seemed working.

**Comment by totoroTree** [comment]
I've done a simple test, using the index of highest confidences as class index, it seems work OK. The details as followed:
1. Expand flags as multiple flags before build_forward_backward
2. The output confidence has shape 300*N, here N is just the number of classes, then find the index of the highest confidence is just the class index

**Comment by RomanSteinberg** [comment]
@placentian , @totoroTree verify at my pull request #219 . Is it what you are describing here?

BTW, now I think that p2 from starting post is not necessary. The loss function `sparse_softmax_cross_entropy_with_logits()` working well with data I provided. I mean there were no errors. I have very poor classification results, but I am not sure it is connected.
